protocol/vm/vm.go

```diff
@@ -47,94 +40,42 @@ var ErrFalseVMResult = errors.New("false VM result")
 // execution.
 var TraceOut io.Writer
 
-func VerifyTxInput(tx *bc.Tx, inputIndex uint32) (err error) {
+func Verify(context *Context) (err error) {
 	defer func() {
-		if panErr := recover(); panErr != nil {
-			err = ErrUnexpected
-		}
-	}()
-	return verifyTxInput(tx, inputIndex)
-}
-
-func verifyTxInput(tx *bc.Tx, inputIndex uint32) error {
-	if inputIndex < 0 || inputIndex >= uint32(len(tx.Inputs)) {
-		return ErrBadValue
-	}
-
-	txinput := tx.Inputs[inputIndex]
-
-	expansionReserved := tx.Version == 1
-
-	f := func(vmversion uint64, prog []byte, args [][]byte) error {
-		if vmversion != 1 {
-			return ErrUnsupportedVM
-		}
-
-		vm := virtualMachine{
-			tx:         tx,
-			txContext:  *tx.VMContexts[inputIndex],
-			inputIndex: inputIndex,
-
-			expansionReserved: expansionReserved,
-
-			mainprog: prog,
-			program:  prog,
-			runLimit: initialRunLimit,
-		}
-		for _, arg := range args {
-			err := vm.push(arg, false)
-			if err != nil {
-				return err
+		if r := recover(); r != nil {
+			if rErr, ok := r.(error); ok {
+				err = errors.Sub(ErrUnexpected, rErr)
```

Comment:
This looks like it's doing some new work to preserve error information if the panic value is an error. Is this really related to switching the dependency direction? I don't understand how. I think it's a good change, but maybe in its own PR?